board: Add support for the Arduino Nano Matter, fix BLE on the SparkFun ThingPlus Matter

silabs-bozont opened the pull request:
This PR adds support for the **Arduino Nano Matter** (MGM240SD22VNA) and Silicon Labs MGM modules.
Silicon Labs radio modules come pre-certified and need special variants of the radio abstraction (RAIL) library along with their respective radio configurations.

This PR also fixes BLE on the **SparkFun Thing Plus Matter** (MGM240PB32VNA) by selecting the correct radio library and config files.

Device tree for the modules mentioned above were created with their proper configurations.
The cmake environment was also modified to take care of linking with the correct module archives.

Depends on: https://github.com/zephyrproject-rtos/hal_silabs/pull/109

silabs-bozont committed:
soc: silabs: Add support for the MGM240SD22VNA

Also introduce the framework to support other
Silicon Labs modules.

Signed-off-by: Tamas Jozsi <tamas.jozsi@silabs.com>

silabs-bozont committed:
boards: fix Bluetooth LE support on the SparkFun ThingPlus Matter MGM240

Create and select the proper module device tree file which loads the
correct radio config for the MGM240P module.

Signed-off-by: Tamas Jozsi <tamas.jozsi@silabs.com>

jerome-pouiller commented:
You have to change the reference to `hal_silabs` in `west.yml` in `pull/109/head`. Otherwise, the CI will probably fail.
(Once the CI will be green, PR 109 in `hal_slabs` will be merged and you will need to update `west.yml` with the final commit id and this PR will be merged)

jerome-pouiller reviewed (commented):
Split this commit. Probably you want to introduce MGM240SD22VNA, then introduce `nano_matter`.

If you fix BLE, keep this change in a separated commit.

silabs-bozont committed:
boards: arduino: Add support for the Arduino Nano Matter

Add device tree and support files for the Nano Matter board
based on the Silicon Labs MGM240SD22VNA chip.

Signed-off-by: Tamas Jozsi <tamas.jozsi@silabs.com>

nordicjm commented on boards/arduino/nano_matter/arduino_nano_matter.dts:
```diff
@@ -0,0 +1,223 @@
+			gpios = <&gpioc 1 GPIO_ACTIVE_LOW>;
+			label = "LED 0";
+		};
+		led1: led_1 {
```
newline gap between child nodes

nordicjm commented on west.yml:
```diff
@@ -235,7 +235,7 @@ manifest:
         - hal
     - name: hal_silabs
-      revision: 95fb8af40b1fd6d9c9cb5ed99ef2f60930361b2b
+      revision: pull/109/head
```
if this is needed by the previous commit then this commit should come first

kartben reviewed (dismissed):
+1 for docs - thanks!

asmellby commented on modules/hal_silabs/simplicity_sdk/CMakeLists.txt:
```diff
@@ -126,13 +157,19 @@ zephyr_include_directories(
 
+if(SILABS_DEVICE_IS_MODULE)
+  zephyr_compile_definitions(
+  _SILICON_LABS_MODULE=1
```
Why is this needed? This macro is provided through `em_device.h`, included by `soc.h` in Zephyr.

silabs-bozont replied:
`modules/hal/silabs/simplicity_sdk/platform/radio/rail_lib/plugin/pa-conversions/pa_curves_efr32.c` checks this macro without including `em_device.h` first - and it causes the default SoC PA Curves to be included for modules as well. I added this here to prevent the inclusion of the SoC Curves while using modules.

asmellby commented on modules/hal_silabs/simplicity_sdk/CMakeLists.txt:
```diff
@@ -23,8 +23,15 @@ set(BLOBS_DIR      ${ZEPHYR_HAL_SILABS_MODULE_DIR}/zephyr/blobs/simplicity_sdk)
-# Get SoC series number, i.e. translate e.g. efr32bg22 -> 22
-string(SUBSTRING ${CONFIG_SOC_SERIES} 7 2 SILABS_DEVICE_FAMILY_NUMBER)
+if("${CONFIG_SOC_PART_NUMBER}" MATCHES "MGM24")
+  set(SILABS_DEVICE_IS_MODULE TRUE)
```
Please introduce a new Kconfig symbol for this, and set it for the MGM240 series, rather than hard-coding it in CMake.

asmellby commented on modules/hal_silabs/simplicity_sdk/CMakeLists.txt:
```diff
@@ -23,8 +23,15 @@ set(BLOBS_DIR      ${ZEPHYR_HAL_SILABS_MODULE_DIR}/zephyr/blobs/simplicity_sdk)
-string(SUBSTRING ${CONFIG_SOC_SERIES} 7 2 SILABS_DEVICE_FAMILY_NUMBER)
+if("${CONFIG_SOC_PART_NUMBER}" MATCHES "MGM24")
+  set(SILABS_DEVICE_IS_MODULE TRUE)
+  set(SILABS_DEVICE_FAMILY "MGM24")
```
This option and the two below should all be derived from SOC_* Kconfig options, not hard-coded in CMake.

asmellby commented on modules/hal_silabs/simplicity_sdk/CMakeLists.txt:
```diff
@@ -77,8 +99,17 @@ if(CONFIG_SOC_GECKO_HAS_RADIO)
+    # RAIL lib and config for modules
+    if("${CONFIG_SOC_PART_NUMBER}" STREQUAL "MGM240SD22VNA")
+      add_prebuilt_library(librail platform/radio/rail_lib/autogen/librail_release/librail_module_efr32xg24_gcc_release.a)
+      add_prebuilt_library_with_whole_archive(librail_config platform/radio/rail_lib/autogen/librail_release/librail_config_mgm240sd22vna_gcc.a)
```
This should generically use the `CONFIG_SOC` variable to produce the library path.

jerome-pouiller commented on boards/arduino/nano_matter/doc/index.rst:
```diff
@@ -0,0 +1,116 @@
+Overview
+********
+
+The Nano Matter merges the well-known Arduino way of making complex technology more accessible with the powerful MGM240S from Silicon Labs, to bring Matter closer to the maker world, in one of the smallest form factors in the market.
```
This is the "source file" of the documentation, so I suggest to apply 80 columns rule.

jerome-pouiller commented on boards/arduino/nano_matter/doc/index.rst:
```diff
@@ -0,0 +1,116 @@
+The Arduino Nano Matter contains an SAMD11 with CMSIS-DAP, allowing flashing, debugging, logging, etc. over
+the USB port. Doing so requires a version of OpenOCD that includes support for the flash on the MG24
+MCU. Until those changes are included in stock OpenOCD, the version bundled with Arduino can be
+used, or can be installed from the `OpenOCD Arduino Fork`_. When flashing, debugging, etc. you may
```
Same comment than: https://github.com/zephyrproject-rtos/zephyr/pull/88321#discussion_r2037046606. BTW, maybe the already opened PR is sufficient: https://github.com/zephyrproject-rtos/openocd/pull/70. In this case, just add a comment explaining it is also useful for this board.

silabs-bozont commented on modules/hal_silabs/simplicity_sdk/CMakeLists.txt:
```diff
@@ -140,7 +173,7 @@ zephyr_include_directories(
 
 zephyr_compile_definitions(
-  ${SILABS_DEVICE_PART_NUMBER}
+  ${SILABS_DEVICE_PART_NUMBER}=1
```
It works the same without - removing. I think I added it to match the makefile generated by SLC - but it doesn't make a difference.

silabs-bozont commented on soc/silabs/soc.yml:
```diff
@@ -53,6 +53,8 @@ family:
           - name: efr32mg24b310f1536im48
           - name: efr32mg24b210f1536im48
           - name: efr32mg24b020f1536im40
+          - name: mgm240sd22vna
```
Thanks for the remarks!
I've renamed the `efr32mg24` folder to `xg24` - added a separate config for the MGM24.
Also introduced the config `SOC_SERIES_XG24`which is set for both MG24 and MGM24 - and changed `clock_control_silabs.h` to check this macro instead.

asmellby commented on boards/arduino/nano_matter/doc/index.rst:
```diff
@@ -0,0 +1,121 @@
+System Clock
+============
+
+The MGM240S SiP is configured to use the 39 MHz internal oscillator.
```
Looks like DTS configures the CPU to use 78 MHz from the DPLL?

asmellby commented on soc/silabs/silabs_s2/xg24/Kconfig.soc:
```diff
@@ -23,17 +37,32 @@ config SOC_PART_NUMBER_EFR32MG24B210F1536IM48
+config SOC_PART_NUMBER_MGM240SD22VNA
+	bool
+	select SOC_SERIES_MGM24
+	select SILABS_DEVICE_IS_MODULE
```
I don't believe `Kconfig.soc` should reference symbols outside of the `SOC_` namespace (the `Kconfig.soc` files should be possible to use without including any `Kconfig` files). This should likely be done in the normal `Kconfig` instead (as part of all the selections for `SOC_SERIES_MGM24`).

asmellby commented on modules/hal_silabs/simplicity_sdk/CMakeLists.txt:
```diff
@@ -23,8 +23,13 @@ set(BLOBS_DIR      ${ZEPHYR_HAL_SILABS_MODULE_DIR}/zephyr/blobs/simplicity_sdk)
+  string(SUBSTRING ${CONFIG_SOC_SERIES} 7 2 SILABS_DEVICE_FAMILY_NUMBER)
+elseif(NOT SILABS_DEVICE_FAMILY_NUMBER GREATER_EQUAL "0")
+  string(SUBSTRING ${CONFIG_SOC_SERIES} 4 3 SILABS_DEVICE_FAMILY_NUMBER)
+endif()
```
It would at least need to be `^.gm` to match bgm, zgm, fgm, etc, but it's possible.

jerome-pouiller commented:
Why do we need the `--whole-archive` stuff exactly?

jhedberg commented:
I manually removed the manifest DNM label, since that seems to have been stuck.

jhedberg commented:
> I manually removed the manifest DNM label, since that seems to have been stuck. 

Actually, this wasn't a bug but a feature of the new process, however I verified that the changes are ok: "⛔ DNM label due to: 1 project with metadata changes and 3 blob changes"

silabs-bozont commented:
> Why do we need the `--whole-archive` stuff exactly?
Tested functionality with `add_prebuilt_library` only - works the same - removed the `--whole-archive` stuff.
I added it when the default SoC PA curves were still linked - now it's not needed anymore.

kartben commented on boards/arduino/nano_matter/arduino_nano_matter-pinctrl.dtsi:
```diff
@@ -0,0 +1,60 @@
+&pinctrl {
+	eusart0_default: eusart0_default {
+		group0 {
+			pins = <EUSART1_TX_PA4>;
+			drive-push-pull;
+			output-high;
+		};
+
+		group1 {
+			pins = <EUSART1_RX_PA5>;
```
you prob want `EUSART0_TX_xx` here?

asmellby replied:
Had to do a double take, since RX is correct in group1, but yes, agree it should be EUSART**0** both on the TX and RX side.

kartben replied:
ya the comment is one lines 13-19 actually, but depending on where you are Github might make things more confusing than they need to be :)

kartben commented on boards/arduino/nano_matter/arduino_nano_matter.dts:
```diff
@@ -0,0 +1,213 @@
+		};
+
+		/* Reserve 720 kB for the application in slot 1 */
+		slot1_partition: partition@C4000 {
+			reg = <0x000C0000 0x000B8000>;
```
```suggestion
		slot1_partition: partition@C4000 {
			reg = <0x000C4000 0x000B8000>;
```

kartben replied:
I guess the typo comes from xiao_mg24.dts -- worth fixing there if you have a chance

kartben replied:
also, 0xB8000 is 736 KB, not 720, so the comments would need to be fixed (and for xiao as well :) )

silabs-bozont replied:
I fixed it for the Nano Matter - if it's okay I'll fix it for the XIAO MG24 in a separate PR. I would like to get this one merged without widening the scope.

silabs-bozont replied:
@kartben here's the fix for the XIAO MG24: https://github.com/zephyrproject-rtos/zephyr/pull/91990

kartben commented on modules/hal_silabs/simplicity_sdk/CMakeLists.txt:
```diff
@@ -23,8 +23,16 @@ set(BLOBS_DIR      ${ZEPHYR_HAL_SILABS_MODULE_DIR}/zephyr/blobs/simplicity_sdk)
+elseif(CONFIG_SOC_SERIES MATCHES "^.gm")
+  string(SUBSTRING ${CONFIG_SOC_SERIES} 3 2 SILABS_DEVICE_FAMILY_NUMBER)
+else()
+  error("unknown part")
```
i don't think error exists in cmake?

```suggestion
  message("unknown part")
```
